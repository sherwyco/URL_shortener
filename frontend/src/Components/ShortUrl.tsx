import { ChangeEvent, useEffect, useState } from 'react';
import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TablePagination from '@mui/material/TablePagination';
import TableRow from '@mui/material/TableRow';
import DeleteButton from './DeleteButton';
import { getData } from '../Api/Client';
import { maxWidth } from '@mui/system';

interface Column {
    id: 'originalUrl' | 'shortCode' | 'createdAt' | 'objectId';
    label: string;
    minWidth?: number;
    maxWidth?: number,
    align?: "left" | "right" | "center"
}

const columns: readonly Column[] = [
    { id: 'originalUrl', label: 'Url', maxWidth: 100, align: 'left' },
    { id: 'shortCode', label: 'Code', align: "center" },
    {
        id: 'createdAt',
        label: 'Created At',
        align: "center"

    },
    {
        id: 'objectId', label: 'Delete', align: 'right'
    },

];
interface ApiType {
    id: number;
    original_url: string;
    short_code: string;
    created_at: string
}

interface Data {
    originalUrl: string;
    shortCode: string;
    createdAt: string;
    objectId: number;
}

function createData(
    originalUrl: string,
    shortCode: string,
    createdAt: string,
    objectId: number,
): Data {
    const formatCreatedDate = new Date(createdAt).toLocaleString()

    return { originalUrl: originalUrl, shortCode: shortCode, createdAt: formatCreatedDate, objectId: objectId };
}

const rows = [
    createData('https://drive.google.com/file/d/1RGD4flezyrCULKWl9gMp2v9U5_zs2w21/view', 'IEla5uL', "2023-02-02T00:34:07.906105Z", 1),

];


export default function ShortUrl() {
    const [page, setPage] = useState(1);
    const [rowsPerPage, setRowsPerPage] = useState(10);
    const [data, setData] = useState<Data[]>([])

    const handleChangePage = (event: unknown, newPage: number) => {
        setPage(newPage);
    };

    const handleChangeRowsPerPage = (event: ChangeEvent<HTMLInputElement>) => {
        setRowsPerPage(+event.target.value);
        setPage(1);
    };

    const handleDelete = () => {
        console.log("DELETED!")
    }

    const cleanData = (rawData: []) => {
        const row: Data[] = []
        rawData?.forEach((elem: ApiType) => {
            row.push({
                objectId: elem.id,
                originalUrl: elem.original_url,
                shortCode: elem.short_code,
                createdAt: elem.created_at
            })
        })
        return row;
    }


    useEffect(() => {
        getData("/shortener/", `?per_page=${rowsPerPage}&page=${page}`, {}).then((response: any) => {
            setPage(response.data.links.page)
            setRowsPerPage(response.data.links.per_page)
            setData(cleanData(response.data.results));
        }).catch((error) => {
            console.log(error)
        })
    }, [setData, setRowsPerPage, setPage])

    return (
        <Paper sx={{ width: '100%', overflow: 'hidden' }}>
            <TableContainer sx={{ maxHeight: 440 }}>
                <Table stickyHeader aria-label="sticky table">
                    <TableHead>
                        <TableRow>
                            {columns.map((column) => (
                                <TableCell
                                    key={column.id}
                                    align={column.align}
                                    style={{ minWidth: column.minWidth, maxWidth: column.maxWidth }}
                                >
                                    {column.label}
                                </TableCell>
                            ))}
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {data?.map((row) => {
                            return (
                                <TableRow hover role="checkbox" tabIndex={-1} key={row.shortCode}>
                                    {columns.map((column) => {
                                        const value = row[column.id];
                                        return (
                                            <TableCell key={column.id} align={column.align} sx={{
                                                overflow: 'hidden',
                                                textOverflow: 'ellipsis',
                                                maxWidth: 30,
                                            }}>
                                                {column.id === 'objectId' ? <DeleteButton handleDelete={handleDelete} /> : value}
                                            </TableCell>
                                        );
                                    })}
                                </TableRow>
                            );
                        })}
                    </TableBody>
                </Table>
            </TableContainer>
            <TablePagination
                rowsPerPageOptions={[10, 25, 50]}
                component="div"
                count={rows.length}
                rowsPerPage={rowsPerPage}
                page={page}
                onPageChange={handleChangePage}
                onRowsPerPageChange={handleChangeRowsPerPage}
            />
        </Paper>
    );
}