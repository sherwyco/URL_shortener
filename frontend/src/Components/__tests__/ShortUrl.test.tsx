import renderer from 'react-test-renderer';
import ShortUrl from '../ShortUrl';

it('matches component snapshot', () => {
    const component = renderer.create(<ShortUrl />)
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})