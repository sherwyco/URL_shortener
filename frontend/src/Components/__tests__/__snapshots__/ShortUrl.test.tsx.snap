// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches component snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1fw5qro-MuiPaper-root"
>
  <div
    style={
      {
        "margin": "10px 10px 10px 10px",
        "textAlign": "right",
      }
    }
  >
    <div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Add Row
      </button>
    </div>
  </div>
  <div
    className="MuiTableContainer-root css-ga3o5l-MuiTableContainer-root"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignLeft MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
            scope="col"
            style={
              {
                "minWidth": undefined,
              }
            }
          >
            <span
              className="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
            >
              Url
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
            scope="col"
            style={
              {
                "minWidth": undefined,
              }
            }
          >
            <span
              className="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
            >
              Code
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
            scope="col"
            style={
              {
                "minWidth": undefined,
              }
            }
          >
            <span
              className="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
            >
              Created At
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
            scope="col"
            style={
              {
                "minWidth": undefined,
              }
            }
          >
            <span
              className="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
            >
              Actions
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
  <div
    className="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
    >
      <div
        className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
      />
      <p
        className="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
        id=":r2:"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r2: :r1:"
          className="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
          id=":r1:"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={10}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
      >
        0â€“0 of 0
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Go to previous page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to previous page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
            />
          </svg>
        </button>
        <button
          aria-label="Go to next page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to next page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
